File: react-frontend/src/PokerHandEvaluator.java
import java.util.*;

public class PokerHandEvaluator {

    // Helper to define card ranks in order
    private static final String RANK_ORDER = "23456789TJQKA";

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        boolean loop = true;
        while (loop) {
            // Input player's hand (2 cards)
            System.out.println("Enter your 2 card hand (e.g., 'AS KH'):");
            String[] playerHand = scanner.nextLine().toUpperCase().split(" ");

            // Input river cards (3 to 5 cards)
            System.out.println("Enter 3 to 5 river cards (e.g., '2D 3H 4S'):");
            String[] riverCards = scanner.nextLine().toUpperCase().split(" ");

            if (playerHand.length != 2 || riverCards.length < 3 || riverCards.length > 5) {
                System.out.println("Invalid input. You need exactly 2 hand cards and between 3 and 5 river cards.");
                return;
            }

            // Combine player's hand and river cards
            List<String> fullHand = new ArrayList<>();
            fullHand.addAll(Arrays.asList(playerHand));
            fullHand.addAll(Arrays.asList(riverCards));

            // Evaluate the best hand
            String bestHand = evaluateHand(fullHand);
            System.out.println("Your best hand is: " + bestHand);

            System.out.println("Would you like to try again? (y/n)");
            String response = scanner.nextLine();
            if (!response.equals("y")) {
                loop = false;
            }
        }
    }

    // Function to evaluate the hand strength
    public static String evaluateHand(List<String> hand) {
        List<Character> handRanks = new ArrayList<>();
        List<Character> handSuits = new ArrayList<>();

        // Extract ranks and suits from each card
        for (String card : hand) {
            handRanks.add(card.charAt(0)); // Rank
            handSuits.add(card.charAt(1)); // Suit
        }

        // Count the occurrences of each rank
        Map<Character, Integer> rankCount = new HashMap<>();
        for (char rank : handRanks) {
            rankCount.put(rank, rankCount.getOrDefault(rank, 0) + 1);
        }

        // Check for different hands
        boolean flush = checkFlush(handSuits);
        boolean straight = checkStraight(handRanks);
        boolean fourOfAKind = checkOfAKind(rankCount, 4);
        boolean fullHouse = checkFullHouse(rankCount);
        boolean threeOfAKind = checkOfAKind(rankCount, 3);
        boolean twoPair = checkTwoPair(rankCount);
        boolean onePair = checkOfAKind(rankCount, 2);

        // Evaluate the best possible hand
        if (flush && straight) return "Straight Flush";
        if (fourOfAKind) return "Four of a Kind";
        if (fullHouse) return "Full House";
        if (flush) return "Flush";
        if (straight) return "Straight";
        if (threeOfAKind) return "Three of a Kind";
        if (twoPair) return "Two Pair";
        if (onePair) return "One Pair";

        return "High Card";
    }

    // Function to check for Flush (5 cards of the same suit)
    public static boolean checkFlush(List<Character> suits) {
        Map<Character, Integer> suitCount = new HashMap<>();
        for (char suit : suits) {
            suitCount.put(suit, suitCount.getOrDefault(suit, 0) + 1);
        }
        return suitCount.values().stream().anyMatch(count -> count >= 5);
    }

    // Function to check for Straight (5 consecutive cards)
    public static boolean checkStraight(List<Character> ranks) {
        List<Integer> sortedRanks = new ArrayList<>();

        // Convert ranks to their corresponding indices in RANK_ORDER
        for (char rank : ranks) {
            sortedRanks.add(RANK_ORDER.indexOf(rank));
        }

        // Remove duplicates and sort the ranks
        Set<Integer> rankSet = new HashSet<>(sortedRanks);
        sortedRanks = new ArrayList<>(rankSet);
        Collections.sort(sortedRanks);

        // Check for consecutive sequence
        int consecutive = 1;
        for (int i = 1; i < sortedRanks.size(); i++) {
            if (sortedRanks.get(i) == sortedRanks.get(i - 1) + 1) {
                consecutive++;
            } else {
                consecutive = 1;
            }
            if (consecutive == 5) return true;
        }
        return false;
    }

    // Function to check for Four of a Kind, Three of a Kind, or One Pair
    public static boolean checkOfAKind(Map<Character, Integer> rankCount, int kind) {
        return rankCount.values().stream().anyMatch(count -> count == kind);
    }

    // Function to check for Full House (Three of a Kind + One Pair)
    public static boolean checkFullHouse(Map<Character, Integer> rankCount) {
        boolean hasThree = false, hasTwo = false;
        for (int count : rankCount.values()) {
            if (count == 3) hasThree = true;
            if (count == 2) hasTwo = true;
        }
        return hasThree && hasTwo;
    }

    // Function to check for Two Pair
    public static boolean checkTwoPair(Map<Character, Integer> rankCount) {
        int pairCount = 0;
        for (int count : rankCount.values()) {
            if (count == 2) pairCount++;
        }
        return pairCount >= 2;
    }
}

File: react-frontend/src/PokerHandEvaluator.py
# Helper to define card ranks in order
RANK_ORDER = "23456789TJQKA"

# Function to evaluate the hand strength
def evaluate_hand(player_hand, river_cards):
    full_hand = player_hand + river_cards

    hand_ranks = [card[0] for card in full_hand]  # Get ranks
    hand_suits = [card[1] for card in full_hand]  # Get suits

    # Count the occurrences of each rank
    rank_count = {}
    for rank in hand_ranks:
        rank_count[rank] = rank_count.get(rank, 0) + 1

    # Check for different hands
    flush = check_flush(hand_suits)
    straight = check_straight(hand_ranks)
    four_of_a_kind = check_of_a_kind(rank_count, 4)
    full_house = check_full_house(rank_count)
    three_of_a_kind = check_of_a_kind(rank_count, 3)
    two_pair = check_two_pair(rank_count)
    one_pair = check_of_a_kind(rank_count, 2)

    # Evaluate the best possible hand
    if flush and straight:
        return 'Straight Flush'
    if four_of_a_kind:
        return 'Four of a Kind'
    if full_house:
        return 'Full House'
    if flush:
        return 'Flush'
    if straight:
        return 'Straight'
    if three_of_a_kind:
        return 'Three of a Kind'
    if two_pair:
        return 'Two Pair'
    if one_pair:
        return 'One Pair'
    
    return 'High Card'

# Function to check for flush (5 cards of the same suit)
def check_flush(suits):
    suit_count = {}
    for suit in suits:
        suit_count[suit] = suit_count.get(suit, 0) + 1
    return any(count >= 5 for count in suit_count.values())

# Function to check for straight (5 consecutive cards)
def check_straight(ranks):
    sorted_ranks = sorted(set(RANK_ORDER.index(rank) for rank in ranks))
    consecutive = 1
    for i in range(1, len(sorted_ranks)):
        if sorted_ranks[i] == sorted_ranks[i - 1] + 1:
            consecutive += 1
        else:
            consecutive = 1
        if consecutive == 5:
            return True
    return False

# Function to check for Four of a Kind, Three of a Kind, or One Pair
def check_of_a_kind(rank_count, kind):
    return any(count == kind for count in rank_count.values())

# Function to check for Full House (Three of a Kind + One Pair)
def check_full_house(rank_count):
    counts = list(rank_count.values())
    return 3 in counts and 2 in counts

# Function to check for Two Pair
def check_two_pair(rank_count):
    pairs = [count for count in rank_count.values() if count == 2]
    return len(pairs) >= 2

# Function to read user input and display the result
def evaluate_user_hand():
    player_hand = input("Enter your 2 card hand (e.g., 'AS KH'): ").upper().split()
    river_cards = input("Enter 3 to 5 river cards (e.g., '2D 3H 4S'): ").upper().split()

    if len(player_hand) != 2 or not (3 <= len(river_cards) <= 5):
        print("Invalid input. You need exactly 2 hand cards and between 3 and 5 river cards.")
        return

    result = evaluate_hand(player_hand, river_cards)
    print(f"Your best hand is: {result}")

# Run the function to evaluate the user's hand
evaluate_user_hand()
